src/services/redis/backend.rs

```diff
@@ -0,0 +1,457 @@
+        let from = args.offset().map_or_else(|| 0, |offset| offset as isize);
+        let to = args.size().map_or_else(|| -1, |size| size as isize + from);
+        let mut con = self
+            .client
+            .get_tokio_connection_manager()
+            .await
+            .map_err(|e| new_get_async_con_error(e, Operation::Read, path.as_str()))?;
+        let gr = con
+            .getrange::<&str, Vec<u8>>(path.as_str(), from, to)
+            .await
+            .map_err(|e| new_exec_async_cmd_error(e, Operation::Read, path.as_str()))?;
+        let cur = Cursor::new(gr);
+        let compat = Compat::new(cur);
+        Ok(Box::new(compat) as BytesReader)
+    }
+
+    async fn write(&self, path: &str, args: OpWrite, r: BytesReader) -> Result<u64> {
+        let path = path.to_string();
+
+        let mut con = self
+            .client
+            .get_tokio_connection_manager()
+            .await
+            .map_err(|e| new_get_async_con_error(e, Operation::Write, path.as_str()))?;
+
+        let c_path = build_abs_content_path(self.root.as_str(), path.as_str());
+        let size = args.size();
+        let mut reader = r;
+        // TODO: asynchronous writing
+        let mut buf = vec![];
+        let content_length = reader.read_to_end(&mut buf).await?;
+        con.set(c_path, buf)
+            .await
+            .map_err(|e| new_exec_async_cmd_error(e, Operation::Write, &path))?;
+
+        let m_path = build_abs_meta_path(self.root.as_str(), path.as_str());
+        let mut meta = ObjectMetadata::default();
+        let mode = if path.ends_with('/') {
+            ObjectMode::DIR
+        } else {
+            ObjectMode::FILE
+        };
+        let last_modified = OffsetDateTime::now_utc();
+        meta.set_content_length(content_length as u64);
+        meta.set_mode(mode);
+        meta.set_last_modified(last_modified);
+        let ser = bincode::serialize(&meta)
+            .map_err(|e| new_serialize_metadata_error(e, Operation::Write, path.as_str()))?;
+        con.set(m_path, ser)
+            .await
+            .map_err(|e| new_exec_async_cmd_error(e, Operation::Write, path.as_str()))?;
+
+        Ok(size)
+    }
+
+    async fn stat(&self, path: &str, _args: OpStat) -> Result<ObjectMetadata> {
+        let path = path.to_string();
+        let mut con = self
+            .client
+            .get_tokio_connection_manager()
+            .await
+            .map_err(|e| new_get_async_con_error(e, Operation::Stat, path.as_str()))?;
+
+        let m_path = build_abs_meta_path(self.root.as_str(), path.as_str());
+        let bin: Vec<u8> = con
+            .get(m_path)
+            .await
+            .map_err(|e| new_exec_async_cmd_error(e, Operation::Stat, path.as_str()))?;
+
+        bincode::deserialize(&bin[..])
+            .map_err(|e| new_deserialize_metadata_error(e, Operation::Stat, path.as_str()))
+    }
+
+    async fn delete(&self, path: &str, _args: OpDelete) -> Result<()> {
```

Comment:
For deleting, we just need to delete one object.

If the directory is not empty, just return a `DirNotEmpty` error.

Reply:
DirNotEmpty requires an unstable feature of Rust.

I'll replace it with `anyhow!("Directory is not empty")`